File: solutions/java/src/coffeevendingmachine/IngredientStore.java
package coffeevendingmachine;

import java.util.HashMap;
import java.util.Map;

public class IngredientStore {
    private final Map<String, Integer> inventory = new HashMap<>();

    public synchronized void refill(String ingredient, int quantity) {
        inventory.put(ingredient, inventory.getOrDefault(ingredient, 0) + quantity);
    }

    public boolean hasEnoughIngredient(Map<String, Integer> required) {
        for (Map.Entry<String, Integer> entry : required.entrySet()) {
            String ingredient = entry.getKey();
            int available = inventory.getOrDefault(ingredient, 0);
            if (available < entry.getValue()) {
                return false;
            }
        }
        return true;
    }

    public synchronized void consume(Map<String, Integer> required) {
        // deduct ingredient
        for (Map.Entry<String, Integer> entry : required.entrySet()) {
            String ingredient = entry.getKey();
            inventory.put(ingredient, inventory.get(ingredient) - entry.getValue());
        }
    }

    public synchronized int getLevel(String ingredient) {
        return inventory.getOrDefault(ingredient, 0);
    }

    public synchronized Map<String, Integer> getAllIngredients() {
        return new HashMap<>(inventory);
    }
}

File: solutions/typescript/src/CoffeeVendingMachine/IngredientStore.ts
class IngredientStore {
  private inventory: Map<string, number>;

  constructor() {
    this.inventory = new Map();
  }

  public refill(ingredient: string, quantity: number): void {
    this.inventory.set(
      ingredient,
      (this.inventory.get(ingredient) || 0) + quantity,
    );
  }

  public hasEnoughIngredient(required: Map<string, number>): boolean {
    for (const [ingredient, quantity] of required) {
      if ((this.inventory.get(ingredient) || 0) < quantity) return false;
    }
    return true;
  }

  public consume(required: Map<string, number>): void {
    for (const [ingredient, quantity] of required) {
      const currentStock = this.inventory.get(ingredient) || 0;
      if (currentStock < quantity) {
        throw new Error(`Not enough ${ingredient} to consume`);
      }
      this.inventory.set(ingredient, currentStock - quantity);
    }
  }

  public getLevel(ingredient: string): number {
    return this.inventory.get(ingredient) || 0;
  }

  public getAllIngredients(): Map<string, number> {
    return { ...this.inventory };
  }
}
export default IngredientStore;
